eclipse-collections/src/main/java/org/eclipse/collections/impl/map/mutable/UnifiedMap.java

```diff
@@ -2384,17 +2391,63 @@ protected void copyKeys(Object[] result)
         }
 
         @Override
-        public boolean equals(Object obj)
+        public boolean equals(Object object)
         {
-            if (obj instanceof Set)
+            if (this == object)
+            {
+                return true;
+            }
+
+            if (!(object instanceof Set))
             {
-                Set<?> other = (Set<?>) obj;
-                if (other.size() == this.size())
+                return false;
+            }
+
+            Set<?> other = (Set<?>) object;
+            if (this.size() != other.size())
+            {
+                return false;
+            }
+
+            for (int i = 0; i < UnifiedMap.this.table.length; i += 2)
+            {
+                Object cur = UnifiedMap.this.table[i];
+                if (cur == CHAINED_KEY)
                 {
-                    return this.containsAll(other);
```

Comment:
This doesn't look like a strict optimization to me. `containsAll` assumes `other` has a reasonably efficient iterator -- most collections do. The new implementation assumes `other` has a very efficient (`O(1)`) `contains`, which is not always true, for example with tree based implementations. If `other.contains` is `O(log N)`, then this method went from `O(N)` to `O(N log N)`.

Reply:
It's a good point. I was making these methods more consistent with the regular equals() method, while trying to make test assertions more consistent. It's been this way since we open sourced.

```java
    @Override
    public boolean equals(Object object)
    {
        if (this == object)
        {
            return true;
        }

        if (!(object instanceof Map))
        {
            return false;
        }

        Map<?, ?> other = (Map<?, ?>) object;
        if (this.size() != other.size())
        {
            return false;
        }

        for (int i = 0; i < this.table.length; i += 2)
        {
            Object cur = this.table[i];
            if (cur == CHAINED_KEY)
            {
                if (!this.chainedEquals((Object[]) this.table[i + 1], other))
                {
                    return false;
                }
            }
            else if (cur != null)
            {
                K key = this.nonSentinel(cur);
                V value = (V) this.table[i + 1];
                Object otherValue = other.get(key);
                if (!UnifiedMap.nullSafeEquals(otherValue, value) || (value == null && otherValue == null && !other.containsKey(key)))
                {
                    return false;
                }
            }
        }

        return true;
    }
```

Reply:
`Map.equals` is bit more special, because iterating `other` may produce a lot of garbage via `Entry` object creation.
I've always felt that there was something missing in collections specifications in Java: a way to examine performance behavior at runtime and pick appropriate algorithms based on that. I guess this could be annotations or marker interfaces like `OhOneContains`, etc and maybe interface specific things like `Map.GarbageFreeEntry`...